frontend/src/app/json-editor/page.tsx

```diff
@@ -0,0 +1,603 @@
+    setScript((prev) => {
+      const method: {
+        type: "default" | "alpine";
+        script: string;
+        resources: {
+          cpu: number | null;
+          ram: number | null;
+          hdd: number | null;
+          os: string | null;
+          version: number | null;
+        };
+      } = {
+        type: "default",
+        script: `/${prev.type}/${prev.slug}.sh`,
+        resources: {
+          cpu: null,
+          ram: null,
+          hdd: null,
+          os: null,
+          version: null,
+        },
+      };
+      return {
+        ...prev,
+        install_methods: [...prev.install_methods, method],
+      };
+    });
+  };
+
+  const updateInstallMethod = (
+    index: number,
+    key: keyof Script["install_methods"][number],
+    value: Script["install_methods"][number][keyof Script["install_methods"][number]],
+  ) => {
+    setScript((prev) => {
+      const updatedMethods = prev.install_methods.map((method, i) => {
+        if (i === index) {
+          const updatedMethod = { ...method, [key]: value };
+
+          // Update script path if `type` of the install method changes
+          if (key === "type") {
+            updatedMethod.script =
+              value === "alpine"
+                ? `/${prev.type}/alpine-${prev.slug}.sh`
+                : `/${prev.type}/${prev.slug}.sh`;
+          }
+
+          return updatedMethod;
+        }
+        return method;
+      });
+
+      const updated = {
+        ...prev,
+        install_methods: updatedMethods,
+      };
+
+      const result = scriptSchema.safeParse(updated);
+      setIsValid(result.success);
+      return updated;
+    });
+  };
+
+  const removeInstallMethod = (index: number) => {
+    setScript((prev) => ({
+      ...prev,
+      install_methods: prev.install_methods.filter((_, i) => i !== index),
+    }));
+  };
+
+  const addNote = () => {
+    setScript((prev) => ({
+      ...prev,
+      notes: [...prev.notes, { text: "", type: "" }],
+    }));
+  };
+
+  const updateNote = (
+    index: number,
+    key: keyof Script["notes"][number],
+    value: string,
+  ) => {
+    setScript((prev) => {
+      const updated = {
+        ...prev,
+        notes: prev.notes.map((note, i) =>
+          i === index ? { ...note, [key]: value } : note,
+        ),
+      };
+      const result = scriptSchema.safeParse(updated);
+      setIsValid(result.success);
+      return updated;
+    });
+  };
+
+  const removeNote = (index: number) => {
+    setScript((prev) => ({
+      ...prev,
+      notes: prev.notes.filter((_, i) => i !== index),
+    }));
+  };
+
+  const addCategory = (categoryId: number) => {
+    setScript((prev) => ({
+      ...prev,
+      categories: [...new Set([...prev.categories, categoryId])],
+    }));
+  };
+
+  const removeCategory = (categoryId: number) => {
+    setScript((prev) => ({
+      ...prev,
+      categories: prev.categories.filter((id) => id !== categoryId),
+    }));
+  };
+
+  return (
```

Comment:
This component is very large, maybe separate some form parts into child component. Forexample `categories`, `install methods` and `notes`, some of these parts also have their own update functions so it would be easier to follow